src/lib.rs

```diff
@@ -29,6 +29,39 @@ mod bits;
 mod memcmp;
 
 use memchr::memchr;
+use std::rc::Rc;
+use std::sync::Arc;
+
+/// Needle that can be searched for within a haystack. It allows specialized
+/// searcher implementations for needle sizes known at compile time.
+pub trait Needle: AsRef<[u8]> {
+    /// Set to `Some(<usize>)` if and only if the needle's length is known at compile time.
+    const SIZE: Option<usize>;
+}
+
+impl<const N: usize> Needle for [u8; N] {
+    const SIZE: Option<usize> = Some(N);
+}
+
+impl Needle for Box<[u8]> {
```

Comment:
I thought about it but for `Box<[u8; N]>` I would want to implement it with the same way as `[u8; N]` (defining `const SIZE: Option<usize> = Some(N);`
Moreover `Box<[u8; N]>` does not implement `AsRef<[u8]>` :( Maybe we can implement it for anything that is `Deref<Target = [u8; N]>` or something? Haven't tried it yet though

Reply:
But it probably needs unit test though! I'll add some later today

Reply:
My thought would be to implement something like this which would cause `Box<[u8]>` to have `SIZE = None` but `But<[u8; N]>` to still have `SIZE = Some(N)`.

```rust
impl<N: Needle + ?Sized> Needle for Box<N> {
    const SIZE: Option<usize> = N::SIZE;
}
```

The `AsRef<[u8]>` is annoying, I had this problem when I tried something similar in one of my branches. I ended up removing the `AsRef` bound on `Needle` and adding an `fn as_bytes(&self) -> &[u8]` method to the trait instead. Not sure how you feel about this though?

Reply:
Your proposal sounds like a very good idea! Thank you! I'll try that approach and add tests.
About `AsRef<[u8]>` vs `fn as_bytes(&self) -> &[u8]` I don't have a strong opinion on the matter tbh but if `AsRef` is more problematic than it actually helps then we don't have to use it I guess.
I haven't had time to investigate, but do you think `impl<R, T: AsRef<R>> AsRef<R> for Box<T>` (and friends like `Arc` / `Rc`) be true and therefore, should we ask upstream about it?

Reply:
It looks like `Box` already has an [`impl<T: ?Sized> AsRef<T> for Box<T>`](https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-AsRef%3CT%3E) which would conflict with `impl<R, T: AsRef<R>> AsRef<R> for Box<T>` for some `T: AsRef<Self>` :confused:

Reply:
I went ahead with your proposal of adding a `as_bytes` method to the trait. It seems to be the most flexible approach while giving us more control (over inlining etc) at the same time.